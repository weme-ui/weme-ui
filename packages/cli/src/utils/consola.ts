import { exit } from 'node:process'
import { createConsola } from 'consola'

export const consola = createConsola({
  formatOptions: {
    date: false,
  },
})

export async function confirm(
  prompt: string,
  cb: () => void,
  exitCode: boolean | number = 0,
) {
  const result = await consola.prompt(prompt, {
    type: 'confirm',
    initial: true,
    cancel: 'undefined',
  })

  if (result === undefined && exitCode !== false)
    exit(Number(exitCode))

  if (result)
    cb()
}

export async function textPrompt(
  prompt: string,
  placeholder?: string,
  exitCode: boolean | number = 0,
) {
  const answer = await consola.prompt(prompt, {
    type: 'text',
    placeholder,
    cancel: 'undefined',
  })

  if (answer === undefined && exitCode !== false)
    exit(Number(exitCode))

  return answer
}

export async function selectPrompt(
  prompt: string,
  options: Array<string | { label: string, value: string }>,
  exitCode: boolean | number = 0,
) {
  const answer = await consola.prompt(prompt, {
    type: 'select',
    options,
    cancel: 'undefined',
  })

  if (answer === undefined && exitCode !== false)
    exit(Number(exitCode))

  return answer
}
