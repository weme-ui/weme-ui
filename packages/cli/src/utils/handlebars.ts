import handlebars from 'handlebars'
import { camelCase, kebabCase, pascalCase, titleCase } from 'scule'

handlebars.registerHelper('titleCase', value => titleCase(value))
handlebars.registerHelper('kebabCase', value => kebabCase(value))
handlebars.registerHelper('camelCase', value => camelCase(value))
handlebars.registerHelper('pascalCase', value => pascalCase(value))
handlebars.registerHelper('lowerCase', value => value.toLowerCase())
handlebars.registerHelper('erase', (value, target) => value.replaceAll(target, ''))

export function compile(input: string, data?: Record<string, any>) {
  return handlebars.compile(input)(data)
}
