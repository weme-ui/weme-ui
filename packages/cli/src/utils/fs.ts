import { copyFileSync, existsSync, readFileSync, writeFileSync } from 'node:fs'
import { exit, cwd as processCwd } from 'node:process'
import { mkdirp } from 'mkdirp'
import { dirname, isAbsolute, relative, resolve } from 'pathe'
import { consola } from './consola'
import { compile } from './handlebars'

export const fs = {
  /**
   * Read file content.
   */
  read: (path: string, cwd: string = '', throwError = true) => {
    const filePath = resolve(cwd, path)

    if (!existsSync(filePath)) {
      if (throwError) {
        consola.error(`File \`${relative(cwd, filePath)}\` not found.`)
        exit(1)
      }

      return false
    }

    return readFileSync(filePath, 'utf8')
  },

  /**
   * Read json file content.
   */
  readJson: <T = any>(path: string, cwd: string = '') => {
    const content = fs.read(path, cwd, false)

    if (content) {
      try {
        const json = JSON.parse(content)

        if (!json)
          return undefined

        return json as T
      }
      catch (error) {
        consola.error(error)
        exit(1)
      }
    }
  },

  /**
   * Write file content.
   */
  write: async (options: {
    path: string
    content: string
    cwd?: string
    force?: boolean
  }) => {
    const {
      path,
      content,
      cwd = processCwd(),
      force = false,
    } = options

    const fullPath = isAbsolute(path) ? path : resolve(cwd, path)
    const isExists = existsSync(fullPath)

    if (isExists && !force) {
      const override = await consola.prompt(
        `File \`${relative(cwd, fullPath)}\` already exists. Do you want to override it?`,
        {
          type: 'confirm',
          cancel: 'undefined',
        },
      )

      if (override === undefined)
        exit(0)

      if (!override)
        return
    }

    fs.mkdirs(dirname(fullPath))

    writeFileSync(fullPath, `${content}\n`, {
      encoding: 'utf8',
      flag: 'w',
    })

    if (!force) {
      consola.success(`${isExists ? 'Overwrote' : 'Created'} file \`${relative(cwd, fullPath)}\`.`)
    }
  },

  /**
   * Write templates to dest.
   */
  writeTemplates: async (templates: string[], args: Record<string, any>) => {
    const resolved = templates.map(
      src => ({
        src,
        dest: compile(src.replace('.hbs', ''), args),
      }),
    )

    for (const tpl of resolved) {
      const content = fs.read(resolve(args.resolve.template, args.type, tpl.src)) || ''

      if (!content) {
        consola.error(`Failed to read template \`${tpl}\``)
        continue
      }

      const compiled = compile(content, args)
      if (!compiled) {
        consola.error(`Failed to render template \`${tpl}\``)
        continue
      }

      fs.write({
        path: resolve(args.dest, tpl.dest),
        content: compiled,
        force: args.force,
      })
    }

    return resolved
  },

  /**
   * Copy file from src to dest.
   */
  copy: (src: string, dest: string) => {
    if (existsSync(src) && !existsSync(dest))
      return copyFileSync(src, dest)
  },

  /**
   * Create directories.
   */
  mkdirs: (...paths: string[]) => {
    for (const path of paths) {
      if (!existsSync(path))
        mkdirp.sync(path)
    }
  },
}
