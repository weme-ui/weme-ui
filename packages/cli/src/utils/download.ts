import { existsSync } from 'node:fs'
import { exit } from 'node:process'
import createDebugger from 'debug'
import { downloadTemplate } from 'giget'
import { isAbsolute, resolve } from 'pathe'
import { CACHE_PATH } from '../constants'
import { consola } from './consola'
import { fs } from './fs'

interface History {
  repo: string
  path: string
  timestamp: number
}

const histories: History[] = []
const debug = createDebugger('download')

export async function download(repo: string) {
  const url = resolveRepoUrl(repo)

  if (isAbsolute(url) && existsSync(url))
    return url

  const cached = getHistory(repo)

  if (cached.timestamp > 0 && existsSync(cached.path))
    return cached.path

  try {
    consola.start('Downloading `%s`...', repo)

    const dest = url.slice(url.lastIndexOf(':') + 1, url.length).replace(/\//g, '-')

    const result = await downloadTemplate(url, {
      cwd: CACHE_PATH,
      dir: dest,
      force: true,
    })

    consola.success('Downloaded `%s` to `%s`', repo, result.dir)

    updateHistory({
      repo,
      path: result.dir,
      timestamp: Date.now(),
    })

    return result.dir
  }
  catch (e: any) {
    consola.error(e.message)
    exit(1)
  }
}

function getHistory(repo: string): History {
  if (histories.length === 0) {
    histories.push(...(fs.readJson<History[]>('history.json', CACHE_PATH) || []))
    debug('history', 'init', histories)
  }

  const history = histories.find(
    h => h.repo === repo
      && h.timestamp > Date.now() - 60 * 60 * 1000,
  )

  debug('history', 'get', history)

  return history || {
    repo,
    path: '',
    timestamp: 0,
  }
}

function updateHistory(history: History) {
  const index = histories.findIndex(h => h.repo === history.repo)

  if (index > -1) {
    histories[index] = history
    debug('history', 'update', history)
  }
  else {
    histories.push(history)
    debug('history', 'add', history)
  }

  fs.write({
    path: 'history.json',
    content: JSON.stringify(histories, null, 2),
    cwd: CACHE_PATH,
    force: true,
  })
}

function resolveRepoUrl(repo: string) {
  if (repo.startsWith('http') || repo.startsWith('@')) {
    const aliases = {
      '@': 'gh:',
      'https://github.com/': 'gh:',
    }

    let alias = repo

    Object.entries(aliases)
      .forEach(([key, value]) => {
        if (repo.startsWith(key))
          alias = repo.replace(key, value)
      })

    return alias
  }

  if (repo.includes(':'))
    return repo

  return resolve(repo)
}
