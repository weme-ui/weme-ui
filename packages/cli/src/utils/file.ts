import { copyFileSync, existsSync, readFileSync, writeFileSync } from 'node:fs'
import { cwd, exit } from 'node:process'
import { mkdirp } from 'mkdirp'
import { dirname, relative, resolve } from 'pathe'
import { consola } from '~/utils/consola'

export function readFile(path: string, cwd: string = '', throwError = true) {
  const filePath = resolve(cwd, path)

  if (!existsSync(filePath)) {
    if (throwError)
      consola.error(`File \`${filePath}\` not found.`)

    return false
  }

  return readFileSync(filePath, 'utf8')
}

export async function writeFile(path: string, content: string, options = { flag: 'w', quiet: false }) {
  const {
    flag = 'w',
    quiet = false,
  } = options

  const isExists = existsSync(path)

  if (isExists && !quiet) {
    const override = await consola.prompt(`File \`${relative(cwd(), path)}\` already exists. Do you want to override it?`, {
      type: 'confirm',
      cancel: 'undefined',
    })

    if (override === undefined)
      exit(0)

    if (!override)
      return
  }

  mkdirs(dirname(path))

  writeFileSync(path, content, {
    encoding: 'utf8',
    flag,
  })

  if (!quiet) {
    consola.info(`${isExists ? 'Overwrote' : 'Created'} file \`${relative(cwd(), path)}\``)
  }
}

export function copyFile(src: string, dest: string) {
  if (existsSync(src) && !existsSync(dest))
    return copyFileSync(src, dest)
}

export function mkdirs(...paths: string[]) {
  for (const path of paths) {
    if (!existsSync(path))
      mkdirp.sync(path)
  }
}
