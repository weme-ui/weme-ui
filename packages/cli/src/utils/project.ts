import { existsSync, readFileSync } from 'node:fs'
import { cwd } from 'node:process'
import { resolve } from 'pathe'

const SCHEMA_FILENAME = 'component.json'

export function isInitialized() {
  return existsSync(resolve(cwd(), `./${SCHEMA_FILENAME}`))
}

export function getSchema() {
  const schemaPath = resolve(cwd(), `./${SCHEMA_FILENAME}`)

  if (!existsSync(schemaPath)) {
    return undefined
  }

  return JSON.parse(readFileSync(schemaPath, 'utf8'))
}

export function hasNuxt() {
  return hasDependency('nuxt')
}

export function hasDependency(name: string) {
  try {
    const packageContent = readFileSync(resolve(cwd(), 'package.json'), 'utf8')
    const packageJSON = packageContent ? JSON.parse(packageContent) : {}
    const { dependencies = {}, devDependencies = {} } = packageJSON

    return !!dependencies[name] || !!devDependencies[name]
  }
  catch {
    return false
  }
}
