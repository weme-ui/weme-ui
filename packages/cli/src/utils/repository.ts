import { existsSync } from 'node:fs'
import { resolve } from 'pathe'
import { existsPackageJson } from './utilities'

export function existsRegistry(name: string, cwd: string) {
  return existsPackageJson(cwd) && existsSync(resolve(cwd, 'registry', name))
}

export function existsComponent(registry: string, name: string, cwd: string) {
  return existsSync(resolve(cwd, 'registry', registry, 'components', name))
}
