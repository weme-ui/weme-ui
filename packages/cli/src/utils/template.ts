import { existsSync } from 'node:fs'
import { cwd, exit } from 'node:process'
import { downloadTemplate } from 'giget'
import handlebars from 'handlebars'
import { relative, resolve } from 'pathe'
import { camelCase, kebabCase, pascalCase, titleCase } from 'scule'
import { consola } from '~/utils/consola'
import { readFile, writeFile } from './file'

const TEMPLATES_PATH = resolve(cwd(), '.templates')

handlebars.registerHelper('titleCase', value => titleCase(value))
handlebars.registerHelper('kebabCase', value => kebabCase(value))
handlebars.registerHelper('camelCase', value => camelCase(value))
handlebars.registerHelper('pascalCase', value => pascalCase(value))
handlebars.registerHelper('lowerCase', value => value.toLowerCase())
handlebars.registerHelper('erase', (value, target) => value.replaceAll(target, ''))

export function compile(input: string, data?: Record<string, any>) {
  return handlebars.compile(input)(data)
}

export function getTemplatePath(template: string) {
  const dest = template.slice(template.lastIndexOf(':') + 1, template.length).replace(/\//g, '-')
  const templatePath = resolve(TEMPLATES_PATH, dest)

  return existsSync(templatePath) ? templatePath : template
}

export async function downloadRemoteTemplate({ template }: Record<string, any>) {
  const dest = template.slice(template.lastIndexOf(':') + 1, template.length).replace(/\//g, '-')
  const templatePath = resolve(TEMPLATES_PATH, dest)

  if (existsSync(templatePath)) {
    consola.info(`Using cached template: \`${relative(cwd(), templatePath)}\``)
    return templatePath
  }

  consola.start('Downloading template...')

  try {
    const result = await downloadTemplate(template, {
      cwd: TEMPLATES_PATH,
      dir: dest,
    })

    consola.success(`Downloaded template: \`${relative(cwd(), result.dir)}\``)

    return result.dir
  }
  catch (error: any) {
    consola.error(error.message)
    exit(1)
  }
}

export async function renderTemplates({
  templates = [],
  data = {},
  ...options
}: {
  src: string
  dest: string
  templates: { src: string, dest: string }[]
  data?: Record<string, any>
}) {
  const resolvedData = data || {}

  for (const { src, dest } of templates) {
    const fileName = compile(dest, resolvedData)
    const filePath = resolve(options.dest, fileName)
    const content = readFile(resolve(options.src, src)) || ''
    const compiled = compile(content, resolvedData)

    if (!compiled) {
      consola.error(`Failed to render template \`${fileName}\``)
      continue
    }

    await writeFile(filePath, compiled, { flag: 'w', quiet: data.force })
  }
}
