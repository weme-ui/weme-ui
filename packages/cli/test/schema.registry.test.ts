import { resolve } from 'pathe'
import { afterEach, describe, expect, it } from 'vitest'
import { fs } from '../src/utils/fs'
import {
  appendRegistryItem,
  appendRegistryItemFiles,
  existsRegistryItem,
  existsRegistryItemFile,
  getRegistryInfo,
  getRegistryItem,
  listRegistries,
  loadRegistryConfig,
} from '../src/utils/schema/registry'

const cwd = resolve(__dirname, 'fixtures')

describe('registry schema', () => {
  afterEach(() => {
    fs.copy(
      './test/fixtures/registry.json',
      './test/fixtures/registry/test/registry.json',
    )
  })

  it('should list registries', async () => {
    expect(await listRegistries(cwd)).toMatchSnapshot()
  })

  it('should load registry config', () => {
    expect(loadRegistryConfig('test', cwd)).toMatchSnapshot()
  })

  it('should get registry info', () => {
    expect(getRegistryInfo('test', cwd)).toMatchSnapshot()
  })

  it('should get registry item', () => {
    expect(getRegistryItem('test', 'icon', cwd)).toMatchSnapshot()
  })

  it('should check registry item exists', () => {
    expect(existsRegistryItem('test', 'icon', cwd)).toBe(true)
    expect(existsRegistryItem('test', 'not-exists', cwd)).toBe(false)
  })

  it('should check registry item file exists', () => {
    expect(existsRegistryItemFile('test', 'icon', 'components/icon/icon.vue', cwd)).toBe(true)
    expect(existsRegistryItemFile('test', 'icon', 'components/icon/not-exists.vue', cwd)).toBe(false)
  })

  it('should append registry item', async () => {
    await appendRegistryItem('test', 'test', 'component', 'general', ['components/test/test.vue'], cwd)
    expect(loadRegistryConfig('test', cwd)).toMatchSnapshot()
  })

  it('should append registry item files', async () => {
    await appendRegistryItemFiles('test', 'icon', ['components/icon/icon-test.vue'], cwd)
    expect(loadRegistryConfig('test', cwd)).toMatchSnapshot()
  })
})
