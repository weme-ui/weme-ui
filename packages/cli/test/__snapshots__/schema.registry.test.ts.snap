// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`registry schema > should append registry item 1`] = `
{
  "access": "public",
  "dependencies": [
    "clsx",
    "tailwind-merge",
    "tailwind-variants",
  ],
  "description": "The standard components for Weme UI.",
  "devDependencies": [
    "@nuxt/icon",
    "@unocss/nuxt",
    "@vueuse/core",
    "@weme-ui/nuxt",
    "@weme-ui/unocss-preset",
    "reka-ui",
    "unocss",
    "vue",
  ],
  "directory": "registry/std",
  "items": [
    {
      "categories": [
        "general",
      ],
      "description": "A component to display any icon from Iconify.",
      "files": [
        {
          "path": "components/icon/icon.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon-box.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.style.ts",
          "type": "style",
        },
        {
          "path": "components/icon/icon-test.vue",
          "type": "component",
        },
      ],
      "name": "icon",
      "title": "Icon",
      "type": "component",
    },
    {
      "categories": [
        "general",
      ],
      "description": "The test component.",
      "files": [
        {
          "path": "components/test/test.vue",
          "type": "component",
        },
      ],
      "name": "test",
      "title": "test",
      "type": "component",
    },
  ],
  "meta": {
    "authors": [
      "Allen Luo <luoyi@mouji.net>",
    ],
    "bugs": "https://github.com/weme-ui/weme-ui/issues",
    "homepage": "https://github.com/weme-ui/weme-ui",
    "repository": "https://github.com/weme-ui/weme-ui",
  },
  "name": "weme-ui/std",
  "prefix": "ui",
  "version": "0.0.0",
}
`;

exports[`registry schema > should append registry item files 1`] = `
{
  "access": "public",
  "dependencies": [
    "clsx",
    "tailwind-merge",
    "tailwind-variants",
  ],
  "description": "The standard components for Weme UI.",
  "devDependencies": [
    "@nuxt/icon",
    "@unocss/nuxt",
    "@vueuse/core",
    "@weme-ui/nuxt",
    "@weme-ui/unocss-preset",
    "reka-ui",
    "unocss",
    "vue",
  ],
  "directory": "registry/std",
  "items": [
    {
      "categories": [
        "general",
      ],
      "description": "A component to display any icon from Iconify.",
      "files": [
        {
          "path": "components/icon/icon.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon-box.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.style.ts",
          "type": "style",
        },
        {
          "path": "components/icon/icon-test.vue",
          "type": "component",
        },
      ],
      "name": "icon",
      "title": "Icon",
      "type": "component",
    },
    {
      "categories": [
        "general",
      ],
      "description": "The test component.",
      "files": [
        {
          "path": "components/test/test.vue",
          "type": "component",
        },
      ],
      "name": "test",
      "title": "test",
      "type": "component",
    },
  ],
  "meta": {
    "authors": [
      "Allen Luo <luoyi@mouji.net>",
    ],
    "bugs": "https://github.com/weme-ui/weme-ui/issues",
    "homepage": "https://github.com/weme-ui/weme-ui",
    "repository": "https://github.com/weme-ui/weme-ui",
  },
  "name": "weme-ui/std",
  "prefix": "ui",
  "version": "0.0.0",
}
`;

exports[`registry schema > should get registry info 1`] = `
{
  "access": "public",
  "authors": [
    "Allen Luo <luoyi@mouji.net>",
  ],
  "bugs": "https://github.com/weme-ui/weme-ui/issues",
  "dependencies": 3,
  "description": "The standard components for Weme UI.",
  "devDependencies": 8,
  "directory": "registry/std",
  "homepage": "https://github.com/weme-ui/weme-ui",
  "items": 2,
  "name": "weme-ui/std",
  "prefix": "ui",
  "repository": "https://github.com/weme-ui/weme-ui",
  "version": "0.0.0",
}
`;

exports[`registry schema > should get registry item 1`] = `
{
  "categories": [
    "general",
  ],
  "description": "A component to display any icon from Iconify.",
  "files": [
    {
      "path": "components/icon/icon.vue",
      "type": "component",
    },
    {
      "path": "components/icon/icon.props.ts",
      "type": "type",
    },
    {
      "path": "components/icon/icon-box.vue",
      "type": "component",
    },
    {
      "path": "components/icon/icon-box.props.ts",
      "type": "type",
    },
    {
      "path": "components/icon/icon-box.style.ts",
      "type": "style",
    },
    {
      "path": "components/icon/icon-test.vue",
      "type": "component",
    },
  ],
  "name": "icon",
  "title": "Icon",
  "type": "component",
}
`;

exports[`registry schema > should list registries 1`] = `
[
  "test",
]
`;

exports[`registry schema > should load registry config 1`] = `
{
  "access": "public",
  "dependencies": [
    "clsx",
    "tailwind-merge",
    "tailwind-variants",
  ],
  "description": "The standard components for Weme UI.",
  "devDependencies": [
    "@nuxt/icon",
    "@unocss/nuxt",
    "@vueuse/core",
    "@weme-ui/nuxt",
    "@weme-ui/unocss-preset",
    "reka-ui",
    "unocss",
    "vue",
  ],
  "directory": "registry/std",
  "items": [
    {
      "categories": [
        "general",
      ],
      "description": "A component to display any icon from Iconify.",
      "files": [
        {
          "path": "components/icon/icon.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.vue",
          "type": "component",
        },
        {
          "path": "components/icon/icon-box.props.ts",
          "type": "type",
        },
        {
          "path": "components/icon/icon-box.style.ts",
          "type": "style",
        },
        {
          "path": "components/icon/icon-test.vue",
          "type": "component",
        },
      ],
      "name": "icon",
      "title": "Icon",
      "type": "component",
    },
    {
      "categories": [
        "general",
      ],
      "description": "The test component.",
      "files": [
        {
          "path": "components/test/test.vue",
          "type": "component",
        },
      ],
      "name": "test",
      "title": "test",
      "type": "component",
    },
  ],
  "meta": {
    "authors": [
      "Allen Luo <luoyi@mouji.net>",
    ],
    "bugs": "https://github.com/weme-ui/weme-ui/issues",
    "homepage": "https://github.com/weme-ui/weme-ui",
    "repository": "https://github.com/weme-ui/weme-ui",
  },
  "name": "weme-ui/std",
  "prefix": "ui",
  "version": "0.0.0",
}
`;
